chainindexer.go

```diff
@@ -0,0 +1,457 @@
+// catch up workers to the database. Since we have concurrent workers, the writer
+// ensures indexes are written in ascending order to avoid a possible gap in the
+// address index triggered by an unexpected shutdown.
+// NOTE: Must be run as a goroutine
+func (a *addrIndexer) indexWriter() {
+	var pendingWrites pendingWriteQueue
+	minHeightWrite := make(chan *writeIndexReq)
+	workerQuit := make(chan struct{})
+	writeFinished := make(chan struct{}, 1)
+
+	// Spawn a goroutine to feed our writer address indexes such
+	// that, if our address tip is at N, the index for block N+1 is always
+	// written first. We use a priority queue to enforce this condition
+	// while accepting new write requests.
+	go func() {
+		for {
+		top:
+			select {
+			case incomingWrite := <-a.writeRequests:
+				heap.Push(&pendingWrites, incomingWrite)
+
+				// Check if we've found a write request that
+				// satisfies our condition. If we have, then
+				// chances are we have some backed up requests
+				// which wouldn't be written until a previous
+				// request showed up. If this is the case we'll
+				// quickly flush our heap of now available in
+				// order writes. We also accept write requests
+				// with a block height *before* the current
+				// index tip, in order to re-index new prior
+				// blocks added to the main chain during a
+				// re-org.
+				writeReq := heap.Pop(&pendingWrites).(*writeIndexReq)
+				_, addrTip, _ := a.server.db.FetchAddrIndexTip()
+				for writeReq.blkHeight == (addrTip+1) ||
+					writeReq.blkHeight <= addrTip {
+					minHeightWrite <- writeReq
+
+					// Wait for write to finish so we get a
+					// fresh view of the addrtip.
+					<-writeFinished
+
+					// Break to grab a new write request
+					if pendingWrites.Len() == 0 {
+						break top
+					}
+
+					writeReq = heap.Pop(&pendingWrites).(*writeIndexReq)
+					_, addrTip, _ = a.server.db.FetchAddrIndexTip()
+				}
+
+				// We haven't found the proper write request yet,
+				// push back onto our heap and wait for the next
+				// request which may be our target write.
+				heap.Push(&pendingWrites, writeReq)
+			case <-workerQuit:
+				return
+			}
+		}
+	}()
+
+out:
+	// Our main writer loop. Here we actually commit the populated address
+	// indexes to the database.
+	for {
+		select {
+		case nextWrite := <-minHeightWrite:
+			err := a.server.db.UpdateAddrIndexForBlock(nextWrite.blkSha,
+				nextWrite.blkHeight, nextWrite.addrIndex)
+			if err != nil {
+				adxrLog.Errorf("Unable to write index for block, "+
+					"sha %v, height %v", nextWrite.blkSha,
+					nextWrite.blkHeight)
+			}
+			writeFinished <- struct{}{}
+		case <-a.quit:
+			break out
+		}
+
+	}
+	close(workerQuit)
+	a.wg.Done()
+}
+
+// indexCatchUpWorker indexes the transactions of previously validated and
+// stored blocks.
+// NOTE: Must be run as a goroutine
+func (a *addrIndexer) indexCatchUpWorker(workChan chan *indexBlockMsg,
+	wg *sync.WaitGroup, quit chan struct{}) {
```

Comment:
The `WaitGroup` still needs to be passed in since graceful exits of the workers are handled by the `indexManager`. This is a separate `WaitGroup` independent of the regular struct-wide WG for the `AddrIndexer` so we can signal our workers to exit after we've caught up and wait for them to finish up their final task(s).